### ISSUE TITLE
unable to set `AssociatePublicIpAddress` for network interface to false

### ISSUE THREAD

**Issue by tuananh** [openshift/machine-api-provider-aws#115]
In corporate context, we often have a SCP (service control policy) to block `AssociatePublicIpAddress` when set to `true` by default.

However, in machine-api-provider-aws, we have this code path

https://github.com/openshift/machine-api-provider-aws/blob/318ae2f87120ae75f4dad89ba0e026cf002a803b/pkg/actuators/machine/instances.go#L364-L389

where as when we create new network interface, we did not explicity set `AssociatePublicIpAddress`. This field, when omited, AWS will use the default value which is `true` as [documented here](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-networkinterface.html#cfn-ec2-launchtemplate-networkinterface-associatepublicipaddress)

Together with the `if aws.BoolValue(machineProviderConfig.PublicIP) {` only check, render us unable to set this value to `false` even if we set `machineProviderConfig.PublicIP` to false.

The fix for this could be

- Default value for network interface changes to `AssociatePublicIpAddress` to `false`.
- Or Add a `else` clause to handle `false` path.

What do you think?

**Comment by tuananh** [comment]
@theobarberbany any advice on how we should proceed? I can open a PR afterward.

**Comment by tuananh** [comment]
we tested with this minimal patch (default to false) and it works well for us

```
diff --git a/pkg/actuators/machine/instances.go b/pkg/actuators/machine/instances.go
index 69a47dc5..6211a523 100644
--- a/pkg/actuators/machine/instances.go
+++ b/pkg/actuators/machine/instances.go
@@ -365,6 +365,7 @@ func launchInstance(machine *machinev1beta1.Machine, machineProviderConfig *mach
                {
                        DeviceIndex: aws.Int64(machineProviderConfig.DeviceIndex),
                        SubnetId:    subnetID,
+                       AssociatePublicIpAddress: aws.Bool(false),
                        Groups:      securityGroupsIDs,
                },
        }
```

**Comment by JoelSpeed** [comment]
Perhaps you linked the wrong thing, but, following the link, the only description I can see of the AssociatePublicIPAddress field is 
> Associates a public IPv4 address with eth0 for a new network interface.
> 
> AWS charges for all public IPv4 addresses, including public IPv4 addresses associated with running instances and Elastic IP addresses. For more information, see the Public IPv4 Address tab on the [Amazon VPC pricing page](https://aws.amazon.com/vpc/pricing/).

There is nothing to say that it defaults to `true` that I can see.

Looking at the [SDK reference](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/), it does mention:
```
If launching into a default subnet, the default value is true.
```
Are you perhaps launching your instance into a default subnet?

**Comment by tuananh** [comment]
I'm pretty sure we do not use the default subnet. we delete the default vpc as part of our account provisioning process .

**Comment by tuananh** [comment]
also i corporate serting, there is usually an SCP required AssociatePublicIpAddress explicitly set to false as well

**Comment by JoelSpeed** [comment]
Hmm, double checking on that logic, I think the issue is just line 374
```
if aws.BoolValue(machineProviderConfig.PublicIP) {
```

If it were instead 
```
if machineProviderConfig.PublicIP != nil {
```
The rest of the logic works just the same, except, whatever the user put in explicitly as true or false into that value, would get passed onto AWS. Meaning anyone not expressing an opinion, would still leverage AWS's default value, whatever they determine that to be.

**Comment by JoelSpeed** [comment]
Appears this is a regression as we used to set the value directly, up until https://github.com/openshift/machine-api-provider-aws/pull/78

**Comment by tuananh** [comment]
yeah that works too. we would like to be able to explicitly set that value
